### PR TITLE
SNOW-1347393: Remove fallback and raise NotImplementedError for GroupBy APIs

### PR DESCRIPTION
**Pull request by sfc-gh-nkrishna**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   SNOW-1347393

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   This PR removes fallback and raises NotImplementedError for GroupBy APIs.

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 1]
```diff
@@ -2471,36 +2468,20 @@ def groupby_ngroups(
...
-            return fallback_ngroups()
 
         query_compiler = get_frame_with_groupby_columns_as_index(
             self, by, level, dropna
         )
 
         if query_compiler is None:
-            return fallback_ngroups()
+            ErrorMessage.not_implemented(
+                "GroupBy.ngroups is not implemented for the given parameter combination"
```
Comment (sfc-gh-nkumar) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:2483]: If we can also tell users which exact parameter is not supported. That would be more actionable information. Not a pupr blocker though so If that leads to much bigger refactoring we can as follow up after PuPr..

Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:2483]: For that, we should need to refactor `check_is_groupby_supported_by_snowflake` and `check_is_aggregation_supported_in_snowflake` which is used in a lot of places. I think it's better to do that in a follow-up PR if necessary given the CHANGELOG/our documentation documents unsupported params.

**Review by sfc-gh-mvashishtha — APPROVED** [review]
LGTM

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 1 above)
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:2483]: We could get some info from the code path without changing the code right? Some errors raised below may specific to some usage cases. Can we make "for the given parameter combination" more concrete without changing code here?

Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:2483]: Oh, I see you already have the group method name included. Then it should be good for now.

**Review by sfc-gh-azhan — APPROVED** [review]
LGTM

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 1 above)
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:2483]: The alternative is something like "`Groupby.{agg_name}` is not implemented for `pd.Grouper`, `axis == 1`, both `by` and `level` are configured, `by` contains any non-pandas hashable labels, or unsupported aggregation parameters."

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:2483]: I like detailed error message.

**Review by sfc-gh-nkumar — APPROVED** [review]

**Pull request merged**